### PR TITLE
Add support for passing array attributes via ffi_call and an example demonstrating the different attribute APIs supported by the FFI

### PR DESCRIPTION
**Pull request by dfm**
The XLA FFI supports [a rich interface for specifying "attributes"](https://openxla.org/xla/custom_call#attributes), but this interface hasn't been widely used from the JAX side. We discovered that passing array attributes wasn't supported by the current front end, so I added support for that and took advantage of this change to add a more complete example of the e2e usage of attributes.

The core JAX change that was needed was to add support for passing `np.array` objects (which aren't hashable!) as keyword arguments to the `ffi_call` primitive. To do this, I wrap any unhashable keyword arguments in the `_HashableByObjectId` helper which creates a hashable object based on the `id`.

"Why not pass these unhashable objects as positional arguments?" you ask. Well, the problem is that we need access to the concrete value of this array _when lowering_, so we can't have them turning into tracers.

I don't expect this to introduce any real caching issues, but if anyone has concerns or other suggestions for a better API, I'd love to hear them!

### CODE REVIEW

**Comment by dfm** [comment]
@gspschmid: Great question - thanks for bringing this up! This is more or less the question I was asking, but I think the specific points you bring up are not a concern here. Let me see if I can lay out my thinking here!

## JIT cache

This change doesn't affect which kinds of arguments can be treated as static arguments to `jax.jit`, and the cache is based on the types of the input arguments, not the Jaxpr. This means that this change doesn't affect cache hits when using `jax.jit`. For example, in the example included with this PR, we call FFI call as follows:

```python
def fun(num: int):
  return jex.ffi.ffi_call(
      "array_attr",
      jax.ShapeDtypeStruct((), np.int32),
      array=np.arange(num, dtype=np.int32),
  )
```

and it's totally fine to `jit` this function:

```python
jitted_fun = jax.jit(fun, static_argnums=(0,))
```

and this will cache as expected:

```python
jitted_fun(5)  # traces here
jitted_fun(5)  # cache hit
jitted_fun(3)  # traces again
```

I note that it's still not possible to pass the `np.array` as a static argument to JIT, and that's disallowed for good reason. If that's the behavior that a user needs: they're going to have to find another way, but that has nothing to do with this change!

## Persistent cache

Another cache that we might be concerned about is the [persistent compilation cache](https://jax.readthedocs.io/en/latest/persistent_compilation_cache.html). This one also isn't a problem because that cache is keyed on the lowered HLO, which doesn't include any references to the object IDs. For example, if we lower the example above:

```python
print(jitted_fun.lower(5).as_text())
```

the HLO is:

```
module @jitted_fun attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() -> (tensor<i32> {jax.result_info = "", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.custom_call @array_attr() {backend_config = "", mhlo.backend_config = {array = array<i32: 0, 1, 2, 3, 4>}, operand_layouts = [], result_layouts = [dense<> : tensor<0xindex>]} : () -> tensor<i32>
    return %0 : tensor<i32>
  }
}
```

where the array has already been materialized as:

```
array<i32: 0, 1, 2, 3, 4>
```

Now, things get a little more complicated if our array has an inexact type (e.g. `float32`). In that case, I could imagine the we could end up with unexpected cache misses depending on numerics, but this is also true for any floating point scalar inputs, which are hashable and unaffected by this change.

## XLA caches

There may be questions to ask about how XLA handles caching with array or dict attributes in the HLO, but that's separate from this PR, which just exposes an API for accessing an existing XLA feature. The JAX-specific hashing hacks are stripped before we ever get to XLA.

## Other caches in JAX core?

LRU caches are used throughout JAX core, and I could certainly imagine this complicating logic _somewhere_, but I haven't thought of anywhere where it would!

## Stepping back

I think it's useful to step back and comment a bit more on the motivation for this change. The `ffi_call` function is really just a simple shim to let users define a lowering rule without having to think about MLIR. (The `ffi_call` primitive doesn't really do anything else besides lowering!) If we weren't using `ffi_call`, we could write our own primitive that has array attributes roughly as follows:

```python
def my_ffi_call_lowering(ctx, *operands, array_size: int):
  return jex.ffi.ffi_lowering("array_attr")(ctx, *operands, array=np.arange(array_size))

my_ffi_call_p = core.Primitive()
# impl, abstract_eval, etc...
mlir.register_lowering(my_ffi_call_p, my_ffi_call_lowering, platform=...)

def fun(*args, array_size: int):
  return my_ffi_call_p.bind(..., array_size=array_size)
```

and everything works! The problem comes from the fact that `ffi_call` can't really have any runtime logic in the lowering rule, so we need to have a way to pass the concrete array through from the binding to the lowering. It's important to make sure we can support such a use case (it was a feature request!), and I think this is a reasonable approach!

File: jax/_src/extend/ffi.py [hunk 1]
```diff
@@ -237,12 +238,26 @@ def ffi_call(
...
+  # wrap them in a hashable object that uses the object's id as the hash. Then,
+  # in the lowering, we unwrap the object and use the original object.
+  hashable_kwargs = {}
+  for k, v in kwargs.items():
+    try:
+      hash(v)
+    except TypeError:
+      hashable_kwargs[k] = _HashableByObjectId(v)
+    else:
+      hashable_kwargs[k] = v
```
Comment (dfm) [inline jax/_src/extend/ffi.py:253]: Note: we could imagine using [`ad_checkpoint.WrapHashably`](https://github.com/jax-ml/jax/blob/20122ff7ce1eb44d266621045821b11e554dafcc/jax/_src/ad_checkpoint.py#L357-L378) here instead of [`api_util._HashableByObjectId`](https://github.com/jax-ml/jax/blob/20122ff7ce1eb44d266621045821b11e554dafcc/jax/_src/api_util.py#L725-L732) since they're roughly equivalent, but I wanted to avoid wrapping arguments that didn't need to be wrapped to keep the Jaxpr clean.

**Comment by gspschmid** [comment]
@dfm Thanks for the elaborate answer, makes sense that this can't affect the caching at the JIT boundary! Do you have a sense of why the values actually need to be hashable? Ultimately `HashableById` makes equality among two wrapped values inexact by erring on the side of "not equal". That's a safe choice in the context of caching, I just couldn't easily gather how the hash and equality come into play here.

**Comment by dfm** [comment]
> Do you have a sense of why the values actually need to be hashable?

You know... this is a great question! And in trying to answer it, I realized that there was a _much_ simpler approach here, which doesn't have these same caveats. I should have figured this out earlier!

The hashability requirement was coming from the fact that I was using [`dispatch.apply_primitive`](https://github.com/jax-ml/jax/blob/bdae9ac72ea1bb15c273089d5273d6ec45c778a0/jax/_src/dispatch.py#L81-L99) as the `impl` rule for the `ffi_call_p` primitive, and that function [caches the jitted call to the primitive](https://github.com/jax-ml/jax/blob/7e2f487ada0e792687cbaa5f93ed6780ba8e306d/jax/_src/dispatch.py#L93).

Having worked that out, I've updated the `ffi_call_p` `impl` rule to use a version of the logic from `dispatch.apply_primitive`, but removing that caching. This will have performance implications if the same `ffi_call` is executed several times in eager mode (because the primitive will be re-lowered each time), but this doesn't affect the behavior of `ffi_call` under JIT.

I'm much happier with this approach! Thanks @gspschmid for asking the straightforward question that I should have asked myself in the first place 🤣

**Comment by dfm** [comment]
It's a bit nasty, but @hawkinsp convinced me that it's probably better to just hash by value (rather than id), since these arguments are small (they're going to end up in the HLO anyways). We special case to handle `np.array` and `dict` arguments, and we could add other cases in the future if necessary.

**Review by hawkinsp — APPROVED** [review]
